### PR TITLE
node pools: add CRUD API

### PR DESCRIPTION
**Pull request by lgfa29**
Implement RPC and HTTP endpoints for basic node pool CRUD operations.

Docs preview:
https://nomad-git-f-node-pools-crud-hashicorp.vercel.app/nomad/api-docs/node-pools

### CODE REVIEW

File: nomad/structs/node_pool.go [hunk 1]
```diff
@@ -133,3 +138,39 @@ func (n *NodePoolSchedulerConfiguration) Validate() error {
...
+type NodePoolUpsertRequest struct {
+	NodePools []*NodePool
+	WriteRequest
+}
+
+// NodePoolDeleteRequest is used to make a request to delete a node pool.
+type NodePoolDeleteRequest struct {
+	Names []string
+	WriteRequest
+}
```
Comment (lgfa29) [inline nomad/structs/node_pool.go:176]: These requests take a list to help synchronizing node pools with mutiregion federation.

File: website/content/api-docs/node-pools.mdx [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+- `Name` `(string: <required>)` - Specifies the node pool to create or update.
+  Must have fewer than 128 characters. Only alphanumeric, `-`, and `_` are
+  allowed.
+
+- `Description` `(string: "")` - Specifies the optional human-readable
+  description of the node pool. Must have fewer than 256 characters.
+
+  - `Meta` `(object: null)` - Optional object with string keys and values of
+    metadata to attach to the node pool. Node pool metadata is not used by
+    Nomad and is intended for use by operators and third party tools.
```
Comment (tgross) [inline website/content/api-docs/node-pools.mdx:152]: extraneous indentation here:

```suggestion
- `Meta` `(object: null)` - Optional object with string keys and values of
  metadata to attach to the node pool. Node pool metadata is not used by
  Nomad and is intended for use by operators and third party tools.
```

File: website/content/api-docs/node-pools.mdx [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+- `prefix` `(string: "")`- Specifies a string to filter node pools based on a
+  name prefix. This is specified as a query string parameter.
+
+- `next_token` `(string: "")` - This endpoint supports paging. The `next_token`
+  parameter accepts a string which identifies the next expected node. This
+  value can be obtained from the `X-Nomad-NextToken` header from the previous
+  response.
+
+- `per_page` `(int: 0)` - Specifies a maximum number of nodes to return for
```
Comment (tgross) [inline website/content/api-docs/node-pools.mdx:37]:
```suggestion
- `per_page` `(int: 0)` - Specifies a maximum number of node pools to return for
```

File: website/content/api-docs/node-pools.mdx [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+| ---------------- | ----------------- |
+| `YES`            | `node_pool:read`  |
+
+### Parameters
+
+- `prefix` `(string: "")`- Specifies a string to filter node pools based on a
+  name prefix. This is specified as a query string parameter.
+
+- `next_token` `(string: "")` - This endpoint supports paging. The `next_token`
+  parameter accepts a string which identifies the next expected node. This
```
Comment (tgross) [inline website/content/api-docs/node-pools.mdx:33]:
```suggestion
  parameter accepts a string which identifies the next expected node pool. This
```

File: command/agent/node_pool_endpoint.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	case "PUT", "POST":
+		return s.nodePoolUpsert(resp, req, poolName)
+	case "DELETE":
+		return s.nodePoolDelete(resp, req, poolName)
+	default:
+		return nil, CodedError(http.StatusMethodNotAllowed, ErrInvalidMethod)
+	}
+}
+
+func (s *HTTPServer) nodePoolList(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
```
Comment (tgross) [inline command/agent/node_pool_endpoint.go:45]: I think we can always use this?

```suggestion
func (s *HTTPServer) nodePoolList(resp http.ResponseWriter, req *http.Request) (any, error) {
```

File: nomad/mock/acl.go [hunk 6]
```diff
@@ -4,11 +4,13 @@
 package mock
 
 import (
+	"bytes"
 	"crypto/rand"
 	"crypto/rsa"
 	"crypto/x509"
 	"encoding/pem"
 	"fmt"
+	"html/template"
```
Comment (tgross) [inline nomad/mock/acl.go:13]: I think?

```suggestion
	"text/template"
```

File: nomad/node_pool_endpoint_test.go [hunk 7]
```diff
@@ -0,0 +1,1112 @@
...
+			MinQueryIndex: 1,
+		},
+	}
+	start := time.Now()
+	var resp structs.NodePoolListResponse
+	err := msgpackrpc.CallWithCodec(codec, "NodePool.List", req, &resp)
+	must.NoError(t, err)
+
+	elapsed := time.Since(start)
+	must.Greater(t, 100*time.Millisecond, elapsed)
```
Comment (tgross) [inline nomad/node_pool_endpoint_test.go:395]: Isn't this racy? The 100ms `UpsertNodePools` timer has already started by the time we call `start := time.Now()`, so it's possible to fall under this if the blocking query returns quickly enough after the watch fires. I don't think the specific timing is actually all that relevant to the test as much as the index is?

**Review by tgross — APPROVED** [review]
LGTM

File: nomad/node_pool_endpoint_test.go (hunk 7 above)
Comment (lgfa29) [reply nomad/node_pool_endpoint_test.go:395]: Hum...true. I guess the time checks for the request being blocked for a while, but if it doesn't than the index check fails anyway. This was mostly copy-past from other similar tests so I didn't think too much about them 😅

File: nomad/mock/acl.go (hunk 6 above)
Comment (lgfa29) [reply nomad/mock/acl.go:13]: Yeah, makes sense. This was auto-imported. Thanks!

File: command/agent/node_pool_endpoint.go (hunk 5 above)
Comment (lgfa29) [reply command/agent/node_pool_endpoint.go:45]: I keep forgetting about it 😬

**Pull request merged**